mmtk/src/julia_scanning.rs

```diff
@@ -358,9 +358,20 @@ pub unsafe fn scan_julia_object<EV: EdgeVisitor<JuliaVMEdge>>(obj: Address, clos
 
 pub unsafe fn mmtk_scan_gcstack<'a, EV: EdgeVisitor<JuliaVMEdge>>(
     ta: *const mmtk_jl_task_t,
-    mut closure: &'a mut EV,
-    mut pclosure: Option<&'a mut EV>,
+    closure: &'a mut EV,
+    pclosure: Option<&'a mut EV>,
 ) {
+    // process transitively pinning stack first
+    let pinning_closure = if let Some(c) = pclosure {
```

Comment:
I would suggest having a separate method `mmtk_scan_tpin_gcstack` or something, and deal with the new stack there. The current code is too confusing: the `tpin_gcstack` is scanned with the normal `closure` when `pclosure` has a value. Creating a new function to scan `tpin_gcstack` and only pass the closure that is needed to the new function.